### PR TITLE
JS and R

### LINKED ISSUE #56: Fun stuff in man pages
out of scope for #33 since not in examples.

https://docs.ropensci.org/writexl/reference/write_xlsx.html

ggalt ?ggalt::show_stateface h/t https://twitter.com/noamross/status/832639711499935744?s=20

https://discuss.ropensci.org/t/searchable-metadata-in-help-files-with-htmlwidgets/1078

### LINKED ISSUE #115: JS in R packages
# JS in R packages

https://cran.r-project.org/web/packages/V8/vignettes/v8_intro.html

https://cran.r-project.org/web/packages/js/vignettes/intro.html

https://www.mail-archive.com/r-package-devel@r-project.org/msg05417.html

https://colinfay.me/node-r-package/ (needs Node and npm installed locally)

# JS in knitr/ R Markdown

## JS literate programming

https://rmarkdown.rstudio.com/authoring_knitr_engines.html%23sql#JavaScript

https://github.com/yihui/knitr/blob/6907b428572c130982f6b3d6c91a164a15b94a30/R/engine.R#L483

https://github.com/ColinFay/bubble#knitr

https://pkg.js4shiny.com/reference/html_document_js.html

View time for knitr and one option of js4shiny, compile time for bubble and one option of js4shiny.

## Different problem

https://community.rstudio.com/t/proper-way-to-do-dependency-ingestion-of-js-for-rmarkdown/64807 (that's a thing you'd distribute in a package)

# Shiny

https://connect.thinkr.fr/js4shinyfieldnotes/read-more.html#shiny-js

https://shiny.rstudio.com/articles/packaging-javascript.html

CDN https://github.com/feddelegrand7/bubblyr/blob/master/R/bubbly.R

Shiny apps as packages https://golemverse.org/

# htmlwidgets

https://cran.r-project.org/web/packages/htmlwidgets/index.html

# JS for R users, how to learn

## JS literate programming

but adds a level of complexity

## Learning materials

https://connect.thinkr.fr/js4shinyfieldnotes/

https://github.com/rstudio-conf-2020/js-for-shiny

## Playground

https://pkg.js4shiny.com/reference/repl.html cf comment below

# Others

https://hackernoon.com/r-and-javascript-execution-libraries-integration-40a30726f295

### PR DESCRIPTION
**Pull request by maelle**
Fix #56 
Fix #115

### CODE REVIEW

**Comment by maelle** [comment]
:wave: @gadenbuie, would you have time to have a look at this? You can even be listed as an author if you want to. Thanks a lot for the comments you already made as an actual JS expert! :pray:

**Comment by gadenbuie** [comment]
Sure @maelle I'd be happy to take a look!

**Comment by gadenbuie** [comment]
@maelle sorry this is taking so long! I've been working on it when I can and will have updates soon, hopefully this weekend.

**Comment by maelle** [comment]
Thanks a ton but no hurry at all, I am now on vacation until August the 17th so won't publish it before the second half of August.

**Comment by maelle** [comment]
I'm back so now whenever you have time to read, I should be more responsive. :slightly_smiling_face:

**Comment by gadenbuie** [comment]
Hey @maelle! I just pushed my local edits/updates into PR #126. Feel free to edit, propose changes, etc. If it makes sense to you, I think we could merge the PR into this branch and then discuss or keep working here.

**Comment by maelle** [comment]
Thanks a ton, I'll merge it to this branch and discuss here as you suggest.

**Comment by maelle** [comment]
Thanks again, this is so well explained!! I have a few comments/questions

* Is the author order ok? I feel that your contribution is bigger now, so you could become the first author. I'd mention you first in the tweet in any case.
* Is it too ambitious to publish the blog post tomorrow Thursday? Otherwise we can wait until next week, no hurry.
* What is a reasonable size for a dependency? 5MB? https://cran.r-project.org/web/packages/policies.html
* I replaced tachyons with webcam because webcam uses JS.
* I'm using Chroma highlighting for that chunk because of https://github.com/r-lib/hugodown/issues/72 :thinking: 
* reactor is for writing R code, correct?

**Comment by gadenbuie** [comment]
> Thanks again, this is so well explained!! I have a few comments/questions
> 
> * Is the author order ok? I feel that your contribution is bigger now, so you could become the first author. I'd mention you first in the tweet in any case.

You're welcome @maelle! The author order is fine the way it is, I'm happy we could collaborate 😄 

> * Is it too ambitious to publish the blog post tomorrow Thursday? Otherwise we can wait until next week, no hurry.

Sounds good to me!

One bit of missing content: I realized that one aspect we didn't cover is how to test packaged JS code. I think we could link to [Chapter 12 Build yourself a safety net | Engineering Production-Grade Shiny Apps](https://engineering-shiny.org/step-secure.html#testing-the-interactive-logic), where some ideas are discussed, in particular `puppeteer` and `shinytest`.

> * What is a reasonable size for a dependency? 5MB? [cran.r-project.org/web/packages/policies.html](https://cran.r-project.org/web/packages/policies.html)

Yes, that's a good link to include. 5MB is *huge* for web dependencies, so it's not something I think about much. If dependencies are that size, package authors would most likely want to use a CDN instead of bundling the deps.

> * I replaced tachyons with webcam because webcam uses JS.
> 
> * I'm using Chroma highlighting for that chunk because of [r-lib/hugodown#72](https://github.com/r-lib/hugodown/issues/72) 🤔

Good call! And sorry about those tabs, I feel like I've been battling tabs vs. spaces in RStudio lately and I missed this one. Would it be easier to convert the tabs to spaces for this chunk?

> * reactor is for writing R code, correct?

Oops! Yes, you're right. I know that I've seen another similar package and I confused the two. I can't find the other one now, so we could remove this paragraph.

File: content/post/2020-07-27-js-r/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+```
+
+JS and R, what a clickbait!
+Come for JS, stay for our posts about [Solaris](/2020/05/14/checking-your-r-package-on-solaris/) and [WinBuilder](/2020/04/01/win-builder/). :wink:
+No matter how strongly you believe in JavaScript being the language of the future (see below), you might still gain from using it in your R practice, be it back-end or front-end.
+
+`r hugodown::embed_tweet("1272595824112029696")`
+
+In this blog post, we shall share a roundup of resources around JavaScript for R package developers.
```
Comment (gadenbuie) [inline content/post/2020-07-27-js-r/index.Rmd:45]:
```suggestion
In this blog post, [Garrick Aden-Buie](https://www.garrickadenbuie.com) and I share a roundup of resources around JavaScript for R package developers.
```

File: content/post/2020-07-27-js-r/index.Rmd [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+> All the smart JavaScript content of this post can be attributed to [Garrick Aden-Buie](https://www.garrickadenbuie.com/). :pray:
```
Comment (gadenbuie) [inline content/post/2020-07-27-js-r/index.Rmd:47]:
```suggestion
```

I appreciate the call-out, but this was very much a collaborative effort and I'd be more comfortable with just a link to my page. :relaxed:

**Comment by maelle** [comment]
Thank you!! I'll work on this again tomorrow morning, so we could publish it next week, I am realizing it'll be easier.

It is even hard to type a tab in RStudio, I don't know how you do it without getting an autocompletion instead! Ah! I ended up using space for the chunk, but that was an interesting bug to discover in any case.

**Comment by maelle** [comment]
Out of curiosity, how did you notice the original issue planning this post btw? (https://github.com/r-hub/blog/issues/115)

**Comment by gadenbuie** [comment]
> Out of curiosity, how did you notice the original issue planning this post btw? (#115)

True story: I very frequently find myself wondering how to do X thing and end up searching `rhub blog X`. 80% of the time this gets me exactly what I need to know. The other 20% the thing I want to do is too niche or you haven't written about it yet. I can't remember what I was looking for this time around, but I couldn't find a blog post about it so I came to the issues to look for a planned post or suggest a new one ... and then saw the JS post! And the rest is history (except for whatever thing I was looking to do, that's completely forgotten 😆)

**Comment by maelle** [comment]
That's a story that I liked reading! I feel very lucky that you saw the JS issue. :grin:

I made a few tweaks to the post based on the convo, and I added a second applause button to see what users of the applause package then do.

Regarding testing, I'm not sure yet where we should introduce it. At the moment, being in bundling JS code or web dependencies, we mention re-using JS code (that's hopefully been tested :grimacing: ). Maybe a new section after web dependencies "Robustness of R packages using JS" that'd mention a) the trade-off of adding one language to the mix (not a topic for web dev and Shiny dev where JS is already there, but if you use V8 in your package, there is one system dependency) b) dependencies checking (just a sentence, or a link to some JS article, how does one trust JS libraries, which is much like background checking CRAN packages I suppose) and CDN choice (probably stick to a popular one, or a CDN you maintain yourself), c) how to test code (with JS tool; the Shiny link; for web documents, not sure how). :thinking:

**Comment by maelle** [comment]
@gadenbuie I added a small subsection about robustness and changed the possible publication date to tomorrow 2020-08-25 :-) (I can snooze it until you have time to have a look, of course)

File: content/post/2020-07-27-js-r/index.Rmd [hunk 3]
```diff
@@ -1,10 +1,10 @@
----
+`---
```
Comment (gadenbuie) [inline content/post/2020-07-27-js-r/index.Rmd:1]: Is this a stray <code>`</code>?

File: content/post/2020-07-27-js-r/index.Rmd [hunk 4]
```diff
@@ -223,3 +232,4 @@ Colin Fay wrote [an experimental Node package for calling R](https://colinfay.me
 In this post we went over some resources useful to R package developers looking to use JavaScript code in the backend or docs of their packages, or to help others use JavaScript dependencies. Do not hesitate to share more links or experience in the comments below!
 
 [^size]: For instance, remember that [for CRAN](https://cran.r-project.org/web/packages/policies.html), "*neither data nor documentation should exceed 5MB*".
+`
```
Comment (gadenbuie) [inline content/post/2020-07-27-js-r/index.Rmd:235]: Also this one?

**Comment by gadenbuie** [comment]
I like the subsection and the publication date sounds good to me!

**Comment by maelle** [comment]
Thanks, so I'll publish this tomorrow! Thanks again for contributing :grinning:

**Comment by gadenbuie** [comment]
Thank you!! This was really fun 😆

**Pull request merged**